src/hotspot/share/classfile/stringTable.cpp

```diff
@@ -323,80 +380,139 @@ void StringTable::update_needs_rehash(bool rehash) {
   }
 }
 
-oop StringTable::do_lookup(const jchar* name, int len, uintx hash) {
+oop StringTable::do_lookup(StringWrapper name, uintx hash) {
   Thread* thread = Thread::current();
-  StringTableLookupJchar lookup(thread, hash, name, len);
   StringTableGet stg(thread);
   bool rehash_warning;
-  _local_table->get(thread, lookup, stg, &rehash_warning);
+
+  switch (name.type) {
+  case StringType::OopStr: {
+    StringTableLookupOop lookup(thread, hash, name.oop_str);
+    _local_table->get(thread, lookup, stg, &rehash_warning);
+    break;
+  }
+  case StringType::UnicodeStr: {
+    StringTableLookupUnicode lookup(thread, hash, name.unicode_str, static_cast<int>(name.length));
+    _local_table->get(thread, lookup, stg, &rehash_warning);
+    break;
+  }
+  case StringType::SymbolStr: {
+    StringTableLookupUTF8 lookup(thread, hash, get_symbol_utf8(name), name.length);
+    _local_table->get(thread, lookup, stg, &rehash_warning);
+    break;
+  }
+  case StringType::UTF8Str: {
+    StringTableLookupUTF8 lookup(thread, hash, name.utf8_str, name.length);
+    _local_table->get(thread, lookup, stg, &rehash_warning);
+    break;
+  }
+  default:
+    ShouldNotReachHere();
+  }
+
   update_needs_rehash(rehash_warning);
   return stg.get_res_oop();
 }
 
+// Converts and allocates to a unicode string and stores the unicode length in len
+const jchar *StringTable::to_unicode(StringWrapper wrapped_str, int &len, TRAPS) {
+  switch (wrapped_str.type) {
+  case StringType::UnicodeStr:
+    return wrapped_str.unicode_str;
+  case StringType::OopStr:
+    return java_lang_String::as_unicode_string(wrapped_str.oop_str(), len, CHECK_NULL);
+  case StringType::SymbolStr: {
+    const char* utf8_str = get_symbol_utf8(wrapped_str);
+    int unicode_length = UTF8::unicode_length(utf8_str, wrapped_str.symbol_str->utf8_length());
+    jchar *chars = NEW_RESOURCE_ARRAY(jchar, unicode_length);
+    UTF8::convert_to_unicode(utf8_str, chars, unicode_length);
+    len = unicode_length;
+    return chars;
+  }
+  case StringType::UTF8Str: {
+    int unicode_length = UTF8::unicode_length(wrapped_str.utf8_str);
+    jchar *chars = NEW_RESOURCE_ARRAY(jchar, unicode_length);
+    UTF8::convert_to_unicode(wrapped_str.utf8_str, chars, unicode_length);
+    len = unicode_length;
+    return chars;
+  }
+  default:
+    ShouldNotReachHere();
+  }
+  return nullptr;
+}
+
+Handle StringTable::to_handle(StringWrapper wrapped_str, TRAPS) {
```

Comment:
`to_handle` is a bit too generic - it's like saying `to_pointer`. You are returning the Handle form of a java.lang.String oop.

Reply:
Renamed this to `handle_from_wrapped_string`, should be clearer and less ambiguous.